bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,185 @@
+                        try {
+                            mergeIfAhead(repo, branch, parent);
+                            progress = true;
+                            parentsIt.remove(); // avoid doing pointless merges
+                        } catch(IOException e) {
+                            log.severe("Merge with " + parent + " failed. Reverting...");
+                            repo.abortMerge();
+                            failedMerges.add(branch + " <- " + parent);
+                            failed = true;
+                        }
+                    }
+                } while(progress && failed);
+
+                if (!failedMerges.isEmpty()) {
+                    throw new IOException("There were failed merges:\n" + failedMerges);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        log.info("Ending topobot run");
+    }
+
+    private static Stream<Branch> dependencies(Repository repo, Hash hash, Path depsFile) throws IOException {
+        return repo.lines(depsFile, hash).map(l -> {
+            var lines = l.stream().filter(s -> !s.isEmpty()).collect(Collectors.toList());
+            if (lines.size() > 1) {
+                throw new IllegalStateException("Multiple non-empty lines in " + depsFile.toString() + ": "
+                        + String.join("\n", lines));
+            }
+            return Stream.of(lines.get(0).split(" ")).map(Branch::new);
+        })
+        .orElse(Stream.of(repo.defaultBranch()));
+    }
+
+    private List<Branch> orderedBranches(Repository repo, Path depsFile) throws IOException {
+        List<Edge> deps = new ArrayList<>();
+        for (var branch : branches) {
+            dependencies(repo, repo.resolve("origin/" + branch.name()).orElseThrow(), depsFile)
+                    .forEach(dep -> deps.add(new Edge(dep, branch)));
+        }
+        var defaultBranch = repo.defaultBranch();
+        return TopologicalSort.sort(deps).stream()
+            .filter(branch -> !branch.equals(defaultBranch))
+            .collect(Collectors.toList());
+    }
+
+    private void mergeIfAhead(Repository repo, Branch branch, Branch parent) throws IOException {
+        var fromHash = repo.resolve(parent.name()).orElseThrow();
+        var oldHead = repo.head();
+        if (!repo.contains(branch, fromHash)) {
+            var isFastForward = repo.isAncestor(oldHead, fromHash);
+            repo.merge(fromHash);
+            if (!isFastForward) {
+                log.info("Merged " + parent + " into " + branch);
+                repo.commit("Automatic merge with " + parent, "duke", "duke@openjdk.org");
+            } else {
+                log.info("Fast forwarded " + branch + " to " + parent);
+            }
+            log.info("merge with " + parent + " succeeded. The following commits will be pushed:\n"
+                    + repo.commits("origin/" + branch.name() + ".." + branch.name()).stream()
```

Comment:
Please use `asList()` or close the `stream()`, otherwise this will leak process id:s

Reply:
You might also want to break this out perhaps a few more lines :smile_cat: